pulp_python/app/tasks/publish.py

```diff
@@ -0,0 +1,187 @@
+import os
+import re
+
+from gettext import gettext as _
+from urllib.parse import urljoin
+
+from celery import shared_task
+from django.core.files import File
+from django.template import Context, Template
+
+from pulpcore.plugin import models
+from pulpcore.plugin.tasking import WorkingDirectory, UserFacingTask
+
+from pulp_python.app import models as python_models
+
+
+log = logging.getLogger(__name__)
+
+simple_index_template = """<!DOCTYPE html>
+<html>
+  <head>
+    <title>Simple Index</title>
+    <meta name="api-version" value="2" />
+  </head>
+  <body>
+    {% for name, canonical_name in projects %}
+    <a href="{{ canonical_name }}">{{ name }}</a>
+    {% endfor %}
+  </body>
+</html>
+"""
+
+
+simple_detail_template = """<!DOCTYPE html>
+<html>
+<head>
+  <title>Links for {{ project_name }}</title>
+  <meta name="api-version" value="2" />
+</head>
+<body>
+    <h1>Links for {{ project_name }}</h1>
+    {% for name, path, md5 in project_packages %}
+    <a href="{{ path }}#md5={{ md5 }}" rel="internal">{{ name }}</a><br/>
+    {% endfor %}
+</body>
+</html>
+"""
+
+
+@shared_task(base=UserFacingTask)
+def publish(publisher_pk, repository_pk):
+    """
+    Use provided publisher to create a Publication based on a RepositoryVersion.
+    Args:
+        publisher_pk (str): Use the publish settings provided by this publisher.
+        repository_pk (str): Create a Publication from the latest version of this Repository.
+    """
+    publisher = python_models.PythonPublisher.objects.get(pk=publisher_pk)
+    repository = models.Repository.objects.get(pk=repository_pk)
+    latest_version = models.RepositoryVersion.latest(repository)
+
+    log.info(_('Publishing: repository={repo}, version={version}, publisher={pub}').format(
+        repo=repository.name,
+        version=latest_version.number,
+        pub=publisher.name
+    ))
+
+    with WorkingDirectory():
+        with models.Publication.create(latest_version, publisher) as publication:
+            write_simple_api(publication)
+
+    log.info(_('Publication: {pk} created').format(pk=publication.pk))
+
+
+def write_simple_api(publication):
+    """
+    Writes metadata mimicking the simple api of PyPI for all python packages in the
+    repository version
+
+    https://wiki.python.org/moin/PyPISimple
+
+    Args:
+        publication (pulpcore.plugin.models.Publication): A publication to generate metadata for
+    """
+
+    os.mkdir('simple')
+    project_names = python_models.PythonPackageContent.objects.order_by('name')\
+        .values_list('name', flat=True).distinct()
+
+    # get all PythonPackageContent from within the repository version, newest to oldest
+    # packages_in_version = python_models.PythonPackageContent.objects.filter(
+    #     pk__in=publication.repository_version.content).order_by('-created')
+
+    # group the python packages by the name of the project
+    #aggregated_packages = itertools.groupby(packages_in_version, lambda x: x.name)
```

Comment:
Commented code can be deleted unless we were to tweak the way we were querying packages to avoid multiple DB queries.